examples/control-vector-generator/pca.hpp

```diff
@@ -0,0 +1,267 @@
+    model.eigenvector = ggml_new_tensor_1d(model.ctx, GGML_TYPE_F32, v_diff_original->ne[1]);
+
+    model.buffer = ggml_backend_alloc_ctx_tensors(model.ctx, model.backend);
+
+    ggml_backend_tensor_set(model.v_diff_original, v_diff_original->data, 0, ggml_nbytes(v_diff_original));
+
+    // no need to load anything into square or square_transpose yet
+
+    // initialize model.eigenvector to random vector
+    std::vector<float> random_vec;
+    std::default_random_engine generator(static_cast<unsigned int>(std::time(0)));
+    std::uniform_real_distribution<float> distribution(0.0, 1.0);
+    for (int i = 0; i < v_diff_original->ne[1]; ++i) {
+        random_vec.push_back(distribution(generator));
+    }
+
+    // we don't normalize it at first but that shouldn't be a problem
+    ggml_backend_tensor_set(model.eigenvector, random_vec.data(), 0, ggml_nbytes(model.eigenvector));
+}
+
+struct ggml_cgraph * square_diff_graph(const pca_model & model) {
+    static size_t buf_size = ggml_tensor_overhead() * GGML_DEFAULT_GRAPH_SIZE + ggml_graph_overhead();
+    static std::vector<uint8_t> buf(buf_size);
+
+    struct ggml_init_params params0 = {
+        /*.mem_size   =*/ buf_size,
+        /*.mem_buffer =*/ buf.data(),
+        /*.no_alloc   =*/ true, // the tensors will be allocated later by ggml_allocr_alloc_graph()
+    };
+    struct ggml_context * ctx0 = ggml_init(params0);
+    struct ggml_cgraph  * gf = ggml_new_graph(ctx0);
+
+    struct ggml_tensor * square = ggml_mul_mat(ctx0, model.v_diff_original, model.v_diff_original);
+    //struct ggml_tensor * square_transpose = ggml_transpose(ctx0, square);
+
+    ggml_build_forward_expand(gf, square);
+
+    ggml_free(ctx0);
+    return gf;
+}
+
+struct ggml_tensor * compute_square(const pca_model & model, ggml_gallocr_t allocr, int n_threads) {
+    struct ggml_cgraph * gf = square_diff_graph(model);
+
+    ggml_gallocr_alloc_graph(allocr, gf);
+
+    if (ggml_backend_is_cpu(model.backend)) {
+        ggml_backend_cpu_set_n_threads(model.backend, n_threads);
+    }
+
+#ifdef GGML_USE_METAL
+    if (ggml_backend_is_metal(model.backend)) {
+        ggml_backend_metal_set_n_cb(model.backend, n_threads);
+    }
+#endif
+
+    ggml_backend_graph_compute(model.backend, gf);
+
+    return gf->nodes[gf->n_nodes - 1];
+}
+
+struct ggml_cgraph * power_iteration_graph(const pca_model & model, float tolerance) {
+    static size_t buf_size = ggml_tensor_overhead() * GGML_DEFAULT_GRAPH_SIZE + ggml_graph_overhead();
+    static std::vector<uint8_t> buf(buf_size);
+
+    struct ggml_init_params params0 = {
+        /*.mem_size   =*/ buf_size,
+        /*.mem_buffer =*/ buf.data(),
+        /*.no_alloc   =*/ true, // the tensors will be allocated later by ggml_allocr_alloc_graph()
+    };
+    struct ggml_context * ctx0 = ggml_init(params0);
+    struct ggml_cgraph  * gf = ggml_new_graph(ctx0);
+
+    struct ggml_tensor * b_tensor = ggml_mul_mat(ctx0, model.square, model.eigenvector);
```

Comment:
I don't know where most of the time is spent in this computation, but matrix-vector multiplications are usually limited by memory bandwidth and it is most likely not worth the cost of copying the data to a GPU to compute it.

Reply:
Interesting, because I recall this section running significantly faster offloaded to my RTX 4090 than it did on CPU. It's been a few days, though, and I might be conflating that with time to compute the square matrix. I'll check when I get home.

However, we do around a thousand iterations of this; does that potentially offset the cost of copying data to GPU once at the beginning? (We do throw data back and forth in the interim to do convergence checks, but we can reduce that.)

Reply:
Yes you are right, if the data is used multiple times it is likely worth it to do it on the GPU. Generally what is not worth is copying data to the GPU and then using it only once to do a matrix-vector multiplication.

Reply:
@christianazinn I had a look deeper in the code. To sum up, what I understand is that (pseudo-code):

```python
square = v_diff_original * v_diff_original^T  # shape of square: [n_embd, n_embd]
eigenvector = random_2d_matrix([n_embd, n_embd])
eigenvector = norm(eigenvector) # optional?
host_old_eigenvector = empty_2d_matrix([n_embd, n_embd])
while True:
  b_tensor = square * eigenvector^T
  b_tensor = norm(b_tensor)
  host_new_eigenvector = b_tensor     # copy GPU --> CPU
  host_old_eigenvector = eigenvector  # copy GPU --> CPU
  # take first row of each tensor, subtract each element and ^2 (calc distance?)
  diff = sum_elements((host_new_eigenvector[0] - host_old_eigenvector[0])^2) 
  if sqrt(diff) < tolerance:
    break
  else:
    eigenvector = host_new_eigenvector
    continue
```

If my pseaudo-code is correct, then what I'm trying to do is:
- Unify the build_graph into one function `build_graph_piter(const pca_model & model, bool calc_square = false, float tolerance)`. If `calc_square` is true, then we add the operation to calc the square
- Do subtraction on GPU and only copy the first row back to CPU
- Maybe do multiple iterations at once to avoid copying the tensor from GPU --> CPU too often (can be configure via a CLI arg for example)

Reply:
Also I'm using 7B model so n_embd = 4096. I'm recently having problem to connect my external GPU so I couldn't tested yet, but seems like mat_mul 2 matrices [4096, 4096] on GPU is still worth it (especially they are FP32). It's true that we can minimize copying GPU <--> CPU to gain some performances here.

Reply:
Yes, this is correct. To your note about normalizing first: we do random distribution(0.0,1.0), but the resulting vector isn't necessarily norm 1: suppose we get a 4096-dimensional vector of all 1s. Then the vector has norm sqrt(4096) = 64 != 1. So we should still normalize at first.

Reply:
I added normalize for random vector. Also, quick question: for the output `b_tensor` is written to the final gguf, this `b_tensor` is always normalized, right?

Reply:
Yes, it needs to be. But because normalization is the last operation we do on it, it should always be.